### ISSUE TITLE
[Feature Request] Support disk quota for docker runtime

### ISSUE THREAD

**Issue by Colstuwjx** [kubernetes/kubernetes#54384]
As docker already support `docker run --storage-opt size=1G xxx` such way to limit container disk usage, it would be better if k8s introduced this feature.

On the other hand, it also would be OK if k8s support pass `--storage-opt` in `containers` spec.
Thanks.

**Comment by andyxning** [comment]
Any updates?

**Comment by M0nsieurChat** [comment]
We would be interested as well ! Service provider with little to no control of what data is stored where by our customers. This would be a nice workaround against disk resource exhaustion of nodes whenever the customer craps outside its persistent volumes.

**Comment by andyxning** [comment]
Just a notification about 'storage-opt size=10g': it only works with overlay2 with backing fs being xfs

**Comment by andyxning** [comment]
Before we add support for configuring `size` command line argument we need to first think about:
* How to specify the value for `size`.
* Where to specify the value for `size`.
* This feature is only available for `overlay2`&&`xfs`. How do we detect this pre-requirement correctly.

Maybe @kubernetes/sig-storage-feature-requests @jingxu97 @vishh can give some advises.

**Comment by BSWANG** [comment]
@andyxning Kubernetes 1.8 add `Local ephemeral storage` resource type. Is it possible to pass `ephemeral-storage`  value to `storage-opt size` when primary partition support `xfs+pquota`.

**Comment by Colstuwjx** [comment]
In my opinion, I think `ephemeral-storage` is not a `quota` for POD, instead, it should be named with `limitation`.

As kubelet would like to evict pod if the ephemeral-storage limit exceeded, and adding `storage-opt` with `docker run` would set container overlay fs quota, and would throws `no space left` error if the quota is exceeded, rather than killing the pod. Currently, I think `ephemeral-storage` is also an option for disk resource limit.

Thanks.

**Comment by NickrenREN** [comment]
>Is it possible to pass ephemeral-storage value to storage-opt size when primary partition support xfs+pquota.

I do not think we need to pass `ephemeral-storage` to `storage-opt`, since `ephemeral-storage` will also limit container overlay usage. but as @Colstuwjx said, we will evict the pod if container usage exceeds the limit.

**Comment by BSWANG** [comment]
@Colstuwjx @NickrenREN Others resource limitations like `cpu`, `memory` is the hard limit, container can't use more than resource above limit value on the system/kernel layer. So those limitations not relate to pod eviction.
Since `ephemeral-storage` not use `disk quota` to implement limitation, the `kubelet` must evict pod when disk usage exceed `ephemeral-storage` limitation. If using `disk quota` to do limitation, applications in containers will got an error of `no space left`, And then applications can do something to decrease disk usage. I think this behavior will more consistent with others limitations.

**Comment by RobertKrawitz** [comment]
The problem with ephemeral storage is even worse than the du latency; it's possible for a pod (attached) to create a file and delete it, but keep the file descriptor around.  Without quota support, any pod or container within it can use as much ephemeral storage as it wants and never get caught doing so.

Remove the `unlink "/tmp/foo"` and the pod gets evicted as it should (but not after chewing through an enormous amount of disk space, if it misses the reaper!)

[diskhog.yaml.txt](https://github.com/kubernetes/kubernetes/files/2224657/diskhog.yaml.txt)

**Comment by wongma7** [comment]
@RobertKrawitz it looks like jingxu97 was working on "Local Ephemeral Storage Capacity Isolation" this but idk the current status https://github.com/kubernetes/features/issues/361 , you may wish to contact her

**Comment by shinebayar-g** [comment]
Any support on this? I think this feature is definitely useful.

**Comment by RobertKrawitz** [comment]
In progress (not this issue per se); see https://github.com/kubernetes/kubernetes/pull/66928

**Comment by Colstuwjx** [comment]
I have been checked the [document](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#monitoring-ephemeral-storage-consumption), and I think, for now, it's focusing on monitoring, not enforcement of quota limit.

From my sight, there is some considerations:

1. it should limit pod storage space, to avoid abusing of storage consuming in multi-tenant case, monitor is the basically thing, limit and enforcement is really needed for our users;

2. we are suffering from [perf issue](https://github.com/kubernetes/kubernetes/issues/83608) about using `du` to monitor, I want the `LocalStorageCapacityIsolationFSQuotaMonitoring` gone beta ASAP, but the `beta criterion` still need to be determined as @dashpole [said](https://github.com/kubernetes/kubernetes/pull/79041#issuecomment-504235085) ;

3. current implement of `LocalStorageCapacityIsolationFSQuotaMonitoring` directly use kernel and file system level features, I think it's okay for `emptyDir` and local storage, but it makes split of CSI and LocalStorage into two lines, I am wondering whether we should add a uniformed `DiskQuota` field, just like CPU and Memory, and let them implement the details;

Thanks @RobertKrawitz great job!

**Comment by pacoxu** [comment]
If I understand correctly, this should be solved by https://github.com/kubernetes/enhancements/issues/1029. Currently, it is an alpha-level feature.

I asked to promote it to beta and after some discussions, we think there are several things to do before promoting it to beta. But for this issue, I think we can close it as resolved. And it should be tracked in  https://github.com/kubernetes/enhancements/issues/1029.

See https://github.com/kubernetes/enhancements/issues/1029#issuecomment-840599150 for more details.

> ### Action Items in 1.22～1.23
> - add some metrics or make more visibility
> - benchmark the feature
> - check e2e testings: e2e evolution
> ### Action Items in 1.24-1.25
> - promote it to beta
